### PR TITLE
Fix dumping on exit

### PR DESCRIPTION
**Pull request by mikiw**
## Usage related changes

- Quick fix for double execution of `dump` function.

## Development related changes

- `test_dumping_on_exit` updated to cover scenario with call after load

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/lint.sh`
- [ ] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing

### CODE REVIEW

**Comment by mikiw** [comment]
This probably is related to introduction of `GunicornServer` and workers:

```
    asyncio.run(state.starknet_wrapper.initialize())

    try:
        print(f" * Listening on http://{args.host}:{args.port}/ (Press CTRL+C to quit)")
        pid = os.getpid()
        GunicornServer(app, args).run()
    except KeyboardInterrupt:
        pass
    finally:
        if args.dump_on == DumpOn.EXIT and os.getpid() != pid:
            state.dumper.dump()
            sys.exit(0)
```

**Comment by mikiw** [comment]
@Solpatium do you have any idea why we have 2 processes running after starting devnet?

image

**Comment by FabijanC** [comment]
Probably due to this line each process has its own "global" state
https://github.com/Shard-Labs/starknet-devnet/blob/40d5b6f75fd80e5adea989e5f45ebb76c5c25cde/starknet_devnet/state.py#L57

File: starknet_devnet/server.py [hunk 1]
```diff
@@ -101,11 +102,12 @@ def main():
 
     try:
         print(f" * Listening on http://{args.host}:{args.port}/ (Press CTRL+C to quit)")
+        pid = os.getpid()
```
Comment (FabijanC) [inline starknet_devnet/server.py:105]: I would call this `main_pid`

File: starknet_devnet/server.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 import sys
 import asyncio
+import os
```
Comment (FabijanC) [inline starknet_devnet/server.py:7]: Can you just sort these three imports alphabetically?

File: starknet_devnet/server.py [hunk 3]
```diff
@@ -101,11 +102,12 @@ def main():
 
     try:
         print(f" * Listening on http://{args.host}:{args.port}/ (Press CTRL+C to quit)")
+        pid = os.getpid()
         GunicornServer(app, args).run()
     except KeyboardInterrupt:
         pass
     finally:
-        if args.dump_on == DumpOn.EXIT:
+        if args.dump_on == DumpOn.EXIT and os.getpid() != pid:
```
Comment (FabijanC) [inline starknet_devnet/server.py:110]: Here I would add a comment saying:
```python
# Dump only if process is worker (not main)
```

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: starknet_devnet/server.py (hunk 2 above)
Comment (mikiw) [reply starknet_devnet/server.py:6]: https://github.com/Shard-Labs/starknet-devnet/pull/337/commits/83d64cdb668ba95ace6aa20eaeebc369f6aae290

File: starknet_devnet/server.py (hunk 3 above)
Comment (mikiw) [reply starknet_devnet/server.py:111]: https://github.com/Shard-Labs/starknet-devnet/pull/337/commits/2c18a620649eacf0521e589c083945bdabe72eff

File: starknet_devnet/server.py (hunk 1 above)
Comment (mikiw) [reply starknet_devnet/server.py]: https://github.com/Shard-Labs/starknet-devnet/pull/337/commits/a350c050ab498a10c7d6decde837da54d002c320

**Comment by mikiw** [comment]
Should we use `worker_exit` in gunicorn and always use 1 worker?

**Comment by Solpatium** [comment]
Having more than 1 worker doesn't make sense at the moment as there is no way of sharing the state between them, right?

**Comment by FabijanC** [comment]
> Having more than 1 worker doesn't make sense at the moment as there is no way of sharing the state between them, right?

Yes. If you don't have objections, I'll merge the PR as it currently is.

**Pull request merged**